hail/src/main/scala/is/hail/expr/LowerArrayToStream.scala

```diff
@@ -17,85 +17,46 @@ import is.hail.expr.types.virtual.{TArray, TStream, TStreamable}
-          case _ => a
-        }
+  private def streamify(node: IR): IR = {
+    node match {
+      case ArraySort(a, l, r, comp) => ArraySort(streamify(a), l, r, boundary(comp))
+      case ArrayFold(a, zero, zn, an, body) => ArrayFold(streamify(a), boundary(zero), zn, an, boundary(body))
+      case ArrayFor(a, n, b) => ArrayFor(streamify(a), n, boundary(b))
+      case ArrayFold2(a, accum, valueName, seq, result) => ArrayFold2(
+        streamify(a),
+        accum.map { case (name, value) => (name, boundary(value)) },
+        valueName, seq.map(boundary), boundary(result))
+      case RunAggScan(a, name, init, seq, res, sig) => RunAggScan(streamify(a), name, boundary(init), boundary(seq), boundary(res), sig)
+      case _: MakeStream | _: StreamRange | _: ReadPartition => node.copy(node.children.map(c => boundary(c.asInstanceOf[IR])))
+      case MakeArray(args, t) => MakeStream(args.map(boundary), TStream(t.elementType, t.required))
+      case ArrayRange(start, stop, step) => StreamRange(boundary(start), boundary(stop), boundary(step))
+      case ArrayZip(childIRs, names, body, behavior) => ArrayZip(childIRs.map(streamify), names, boundary(body), behavior)
+      case ArrayMap(a, n, b) => ArrayMap(streamify(a), n, boundary(b))
+      case ArrayFilter(a, n, b) => ArrayFilter(streamify(a), n, boundary(b))
+      case ArrayFlatMap(a, n, b) => ArrayFlatMap(streamify(a), n, streamify(b))
+      case ArrayScan(a, zero, zn, an, body) => ArrayScan(streamify(a), boundary(zero), zn, an, boundary(body))
+      case ArrayLeftJoinDistinct(l, r, ln, rn, keyf, joinf) =>
+        ArrayLeftJoinDistinct(streamify(l), streamify(r), ln, rn, boundary(keyf), boundary(joinf))
+      case x if x.typ.isInstanceOf[TArray] => ToStream(x)
       case _ =>
-        streamify(streamableNode) match {
-          case ToStream(a) => a
-          case s => ToArray(s)
-        }
+        val newChildren = node.children.map(child => boundary(child.asInstanceOf[IR]))
+        if ((node.children, newChildren).zipped.forall(_ eq _))
+          node
+        else
+          node.copy(newChildren)
     }
-
-    println(s"ending unstreamify with ${r}\n")
-    r
   }
 
   def apply(node: IR): IR = {
     println(s"\n\nStarting LowerArrayToStream with: ${ node }")
-    val r = node match {
-      case ArraySort(a, l, r, comp) => ArraySort(streamify(a), l, r, comp)
-      case ArrayFold(a, zero, zn, an, body) => ArrayFold(streamify(a), zero, zn, an, body)
-      case ArrayFor(a, n, b) => ArrayFor(streamify(a), n, b)
-      case ArrayFold2(a, accum, valueName, seq, result) => ArrayFold2(streamify(a), accum, valueName, seq, result)
-      case x: ApplyIR => apply(x.explicitNode)
-      case RunAggScan(a, name, init, seq, res, sig) => RunAggScan(streamify(a), name, init, seq, res, sig)
-      case _ => unstreamify(Copy(node, Children(node).map { case c: IR => apply(c) }))
-    }
-
-    println(s"ending LowerArrayToStream with: ${ r }\n\n")
-    r
+    streamify(node)
```

Comment:
I like this solution. Previously, relying on Copy and Children, we needed to split some nodes out, like ArrayFold, ArrayFold2 to streamify too many arguments. The other issue was knowing when to unstreamify something. That behavior was hard to follow (sometimes we streamify an IR argument, sometimes we apply a child, and we have ToStream, ToArray in both streamify and unstreamiffy. I wasn't successful however having ToStream and ToArray in just one of the two; needed behavior was different. I'll try this version out first thing in the morning.